### PR TITLE
Recursively compute collection type parameters

### LINKED ISSUE #336: Cannot write a map implementing class which is parametrized by single type
```
public class TestMapMain {
    public static void main(String[] args) {
        Bytes bytes = Bytes.allocateElasticOnHeap();

        final Wire wire = WireType.BINARY.apply(bytes);

        wire.getValueOut().object(new Value(2, "b"));

        final Object object = wire.getValueIn().object();

        System.out.println(object);
    }

    private static class Value extends SelfDescribingMarshallable implements BytesMarshallable {
        private IntMap<String> map = new IntMap<>();

        public Value(int x, String y) {
            map.put(x, y);
        }
    }

    private static class IntMap<T> extends HashMap<Integer, T> {
    }
}
```
will produce exception:
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at net.openhft.chronicle.wire.WireMarshaller$MapFieldAccess.<init>(WireMarshaller.java:1258)
	at net.openhft.chronicle.wire.WireMarshaller$FieldAccess.create(WireMarshaller.java:431)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:517)
	at net.openhft.chronicle.wire.WireMarshaller.of(WireMarshaller.java:87)
	at net.openhft.chronicle.wire.WireMarshaller.lambda$static$0(WireMarshaller.java:51)
	at net.openhft.chronicle.core.ClassLocal.computeValue(ClassLocal.java:54)
	at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:228)
	at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:210)
	at java.base/java.lang.ClassValue.get(ClassValue.java:116)
	at net.openhft.chronicle.wire.Marshallable.writeMarshallable(Marshallable.java:131)
	at net.openhft.chronicle.wire.BinaryWire$FixedBinaryValueOut.marshallable(BinaryWire.java:1855)
	at net.openhft.chronicle.wire.ValueOut.typedMarshallable(ValueOut.java:427)
	at net.openhft.chronicle.wire.ValueOut.object(ValueOut.java:658)
	at feed.replication.TestMapMain.main(TestMapMain.java:21)
```

Wire assumes that map implementing class is always parametrized by `<K, V>`, but it's not uncommon for map implementations to e.g. only parametrize on `<V>`.

### PR DESCRIPTION
**Pull request by alamar**
Fixes #336

### CODE REVIEW

File: src/test/java/net/openhft/chronicle/wire/map/MapCustomTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        assertTrue(result.equals(mapsHolder));
+    }
+
+    private static class MapsHolder<T extends Integer> extends SelfDescribingMarshallable {
+        private IntToStringMap i2sMap = new IntToStringMap();
+        private IntMap<Void, String> iMap = new IntMap<>();
+        private IntSuperMap<String, MapsHolder> isMap = new IntSuperMap<>();
+        private TransformingMap<T, BigDecimal, String> transMap = new TransformingMap<>(BigDecimal::new);
+        private GeneralMap gMap = new GeneralMap();
+        // ClassCastException: net.openhft.chronicle.core.util.ObjectUtils$$Lambda$73/1401132667 cannot be cast to Map
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/wire/map/MapCustomTest.java:33]: commented-out code

File: src/main/java/net/openhft/chronicle/wire/WireMarshaller.java [hunk 2]
```diff
@@ -1250,17 +1244,10 @@ else if (type == SortedMap.class || type == NavigableMap.class)
                 collectionSupplier = TreeMap::new;
             else
                 collectionSupplier = newInstance();
-            Type genericType = field.getGenericType();
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:1253]: Nice deduplication

**Review by JerryShea — APPROVED** [review]
nice

File: src/test/java/net/openhft/chronicle/wire/map/MapCustomTest.java (hunk 1 above)
Comment (alamar) [reply src/test/java/net/openhft/chronicle/wire/map/MapCustomTest.java:33]: It is kept for future extension.

**Pull request merged**